src/peft/tuners/lora/tp_layer.py

```diff
@@ -0,0 +1,155 @@
+import warnings
+from typing import Union
+
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+
+from .config import MegatronConfig
+from .layer import LoraLayer
+
+tensor_parallel = MegatronConfig.tensor_parallel
+
+
+class LoraParalleLayer(LoraLayer):
+    def __init__(self, in_features: int, out_features: int, is_paralle_a: bool = False):
+        LoraLayer.__init__(self, in_features=in_features, out_features=out_features)
+        self.is_paralle_a = is_paralle_a
+
+    def update_layer(self, adapter_name, r, lora_alpha, lora_dropout, init_lora_weights, **kwargs):
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+        self.r[adapter_name] = r
+        self.lora_alpha[adapter_name] = lora_alpha
+        if lora_dropout > 0.0:
+            lora_dropout_layer = nn.Dropout(p=lora_dropout)
+        else:
+            lora_dropout_layer = nn.Identity()
+        self.lora_dropout.update(nn.ModuleDict({adapter_name: lora_dropout_layer}))
+
+        config = kwargs['config']
+        # lora needs to be forced to upgrade to 32-bit precision, otherwise it will overflow
+        config.params_dtype = torch.float32
+        init_method = kwargs.get('init_method', init.xavier_normal_)
+        if r > 0:
+            if self.is_paralle_a:
+                lora_a = tensor_parallel.RowParallelLinear(input_size=self.in_features, output_size=r, bias=False,
+                                                           input_is_parallel=kwargs.get('input_is_parallel', True),
+                                                           skip_bias_add=True, init_method=init_method,
+                                                           config=config)
+                lora_b = nn.Linear(in_features=r, out_features=self.out_features, bias=False, dtype=torch.float32)
+            else:
+                lora_a = nn.Linear(in_features=self.in_features, out_features=r, bias=False, dtype=torch.float32)
+                lora_b = tensor_parallel.ColumnParallelLinear(input_size=r, output_size=self.out_features, bias=False,
+                                                              gather_output=kwargs.get('gather_output', False),
```

Comment:
why can't  use lora_a and lora_b both tensor_parallel

Reply:
> why can't use lora_a and lora_b both tensor_parallel

If the target module is RowParallelLinear, the input maybe is parallel, so the lora_a must be RowParallelLinear. At this time, the lora_a's output is complete matrix, although lora_b can be linear or tensor_parallel, but if use tensor_parallel, we will add one more scatter op. Because of lora matrix is small, the efficiency may not necessarily be improved.

And if the target module is ColumnParallelLinear, unless we judge the output is or not gather_out, the most convenient way is lora_b is ColumnParallelLinear, and then the lora_a can be linear or tensor_parallel. But as the first reason, lora matrix is small, if we use linear, the more copy in gpu has little influence, but if we use tensor_parallel, we will add one more scatter op, the communication overhead is big, so the efficiency may not necessarily be improved.